modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/filters/SelectionFilter.js

```diff
@@ -92,73 +93,111 @@ function SelectionFilter({
 	id,
 	inputPlaceholder,
 	itemKey,
-	itemLabel: itemLabelProp,
+	itemLabel,
 	items: initialItems,
 	multiple,
 	selectedData,
 	setFilter,
 }) {
-	const [query, setQuery] = useState('');
-	const [search, setSearch] = useState('');
+	const [searchOptions, setSearchOptions] = useState({
+		currentPage: 1,
+		query: '',
+	});
 	const [selectedItems, setSelectedItems] = useState(
 		selectedData?.selectedItems || []
 	);
 	const [items, setItems] = useState(apiURL ? null : initialItems);
+	const [localItems, setLocalItems] = useState(
+		initialItems.length ? initialItems : []
+	);
 	const [loading, setLoading] = useState(false);
-	const [currentPage, setCurrentPage] = useState(1);
 	const [total, setTotal] = useState(apiURL ? 0 : initialItems?.length);
 	const scrollingAreaRef = useRef(null);
 	const [scrollingAreaRendered, setScrollingAreaRendered] = useState(false);
 	const infiniteLoaderRef = useRef(null);
 	const [infiniteLoaderRendered, setInfiniteLoaderRendered] = useState(false);
 	const [exclude, setExclude] = useState(!!selectedData?.exclude);
 
-	const loaderVisible = items?.length < total;
+	const loaderVisible = !localItems.length && items?.length < total;
 
 	useEffect(() => {
 		setSelectedItems(selectedData?.selectedItems || []);
 	}, [selectedData]);
 
-	useEffect(() => {
-		if (query === search) {
-			return;
-		}
-
-		setCurrentPage(1);
-
-		setSearch(query);
-	}, [query, search]);
+	const handleAutocompleteQuery = debounce((value) => {
+		setSearchOptions({currentPage: 1, search: value});
+	}, DEFAULT_DEBOUNCE_DELAY);
 
 	const isMounted = useIsMounted();
 
+	const firstRequestRef = useRef(true);
+
 	useEffect(() => {
-		if (apiURL) {
+		if (apiURL && !localItems.length) {
 			setLoading(true);
 
-			fetchData(apiURL, search, currentPage)
-				.then((data) => {
+			fetchData(apiURL, searchOptions.query, searchOptions.currentPage)
+				.then((response) => {
+					const selectionItems = response.items.map((item) => {
+						return {
+							label: itemLabel
+								? getValueFromItem(item, itemLabel)
+								: item.label,
+							value: itemKey ? item[itemKey] : item.value,
+						};
+					});
+
 					if (!isMounted()) {
 						return;
 					}
 
 					setLoading(false);
 
-					if (currentPage === 1) {
-						setItems(data.items);
+					if (searchOptions.currentPage === 1) {
+						setItems(selectionItems);
 					}
 					else {
-						setItems((items) => [...items, ...data.items]);
+						setItems((items) => [...items, ...selectionItems]);
+					}
+
+					if (
+						firstRequestRef.current &&
+						response.totalCount <= DEFAULT_PAGE_SIZE &&
+						autocompleteEnabled
+					) {
+						setLocalItems(selectionItems);
 					}
 
-					setTotal(data.totalCount);
+					setTotal(response.totalCount);
+
+					firstRequestRef.current = false;
```

Comment:
Using `firstRequestRef` is enough to enable local search and prevent updating the state re-render the component.